### ISSUE TITLE
Lock is not released when FLoM client receives SIGTERM signal

### ISSUE THREAD

**Issue by karelklic** [tiian/flom#14]
Hi Christian,
when FLoM client running a subprocess receives a SIGTERM signal, it terminates itself without releasing the lock on the FLoM server or telling the subprocess about the signal.

This behavior causes problems is when FLoM runs within the Jenkins automation server. When Jenkins kills a job, it sends SIGTERM to its subprocesses. The FLoM client is terminated immediately, but its lock stays held on the FLoM server shared by all Jenkins jobs. I think the client's subprocess also keeps running.

Below is an unpolished patch that solves this issue. It uses code from https://stackoverflow.com/questions/40477988/calling-kill-on-a-child-process-with-sigterm-terminates-parent-process-but-call. The idea is that signals received by FLoM client are redirected to the subprocess instead of the default behavior. (For SIGTERM, the default behavior is to terminate immediately.) When the subprocess terminates after receiving SIGTERM, the client properly releases the lock and finishes running as well.

```patch
--- a/src/flom_exec.c
+++ b/src/flom_exec.c
@@ -26,7 +26,10 @@
 #ifdef HAVE_SYS_WAIT_H
 # include <sys/wait.h>
 #endif
-
+#include <signal.h>
+#include <string.h>
+#include <stdio.h>
+#include <errno.h>
 
 
 #include "flom_errors.h"
@@ -41,6 +44,44 @@
 #endif /* FLOM_TRACE_MODULE */
 #define FLOM_TRACE_MODULE   FLOM_TRACE_MOD_EXEC
 
+/* Child process PID, and atomic functions to get and set it.
+ * Do not access the internal_child_pid, except using the set_ and get_ functions.
+*/
+static pid_t internal_child_pid = 0;
+
+static inline void set_child_pid(pid_t p)
+{
+  __atomic_store_n(&internal_child_pid, p, __ATOMIC_SEQ_CST);
+}
+
+static inline pid_t get_child_pid(void)
+{
+  return __atomic_load_n(&internal_child_pid, __ATOMIC_SEQ_CST);
+}
+
+static void forward_handler(int signum, siginfo_t *info, void *context)
+{
+    const pid_t target = get_child_pid();
+
+    if (target != 0 && info->si_pid != target)
+        kill(target, signum);
+}
+
+static int forward_signal(const int signum)
+{
+    struct sigaction act;
+
+    memset(&act, 0, sizeof act);
+    sigemptyset(&act.sa_mask);
+    act.sa_sigaction = forward_handler;
+    act.sa_flags = SA_SIGINFO | SA_RESTART;
+
+    if (sigaction(signum, &act, NULL))
+        return errno;
+
+    return 0;
+}
+
 
 int flom_exec(gchar **const command_argv, const char *element,
               int *child_status)
@@ -58,6 +99,16 @@ int flom_exec(gchar **const command_argv, const char *element,
     
     FLOM_TRACE(("flom_exec\n"));
     TRY {
+        /* Install signal forwarders. */
+        if (forward_signal(SIGINT) ||
+            forward_signal(SIGHUP) ||
+            forward_signal(SIGTERM) ||
+            forward_signal(SIGQUIT) ||
+            forward_signal(SIGUSR1) ||
+            forward_signal(SIGUSR2)) {
+            fprintf(stderr, "Cannot install signal handlers: %s.\n", strerror(errno));
+            THROW(COMMAND_ARGV_IS_NULL);
+        }
 
         if (NULL == command_argv) {
             FLOM_TRACE(("flom_exec: command_argv cannot be NULL\n"));
@@ -109,13 +160,36 @@ int flom_exec(gchar **const command_argv, const char *element,
             /* father process */
             FLOM_TRACE(("flom_exec-father: child pid=" PID_T_FORMAT "\n",
                         pid));
+            /* Parent process. Ensure signals are reflected. */
+            set_child_pid(pid);
+
             /* waiting child termination */
             FLOM_TRACE(("flom_exec-father: waiting child termination\n"));
-            if (-1 == (child_pid = wait(&status)))
+            /* Wait until the child we created exits. */
+            while (1) {
+              status = 0;
+              child_pid = waitpid(pid, &status, 0);
+
+              /* Error? */
+              if (child_pid == -1) {
+                /* EINTR is not an error. Occurs more often if
+                   SA_RESTART is not specified in sigaction flags. */
+                if (errno == EINTR)
+                  continue;
+
+                fprintf(stderr, "Error waiting for child to exit: %s.\n", strerror(errno));
                 THROW(WAIT_ERROR);
-            *child_status = WEXITSTATUS(status);
-            FLOM_TRACE(("flom_exec-father: child exit status is %d\n",
-                        *child_status));
+                break;
+              }
+
+              /* Child p exited? */
+              if (child_pid == pid) {
+                *child_status = WEXITSTATUS(status);
+                FLOM_TRACE(("flom_exec-father: child exit status is %d\n",
+                            *child_status));
+                break;
+              }
+            }
         }
         
         THROW(NONE);
```

I use FLoM with this patch for 1.5 years without any issues, although the signal forwarders should perhaps be uninstalled when the subprocess finishes.

**Comment by tiian** [comment]
First of all: many thanks for the interesting topic. I have read it and I think it deserves some serious thinking; moreover, I'm busy developing a complex part of the LIXA project and it will take some time to me to tackle this issue.

The description contains two distinct items:

- FLoM does not redirect signals to the child process

- when flom client terminates, the lock is not released

The first statement is 100% true: I have never thought about it. It deserves a serious analysis and before putting on the table a sloppy solution I need to clarify my ideas about it; for sure I have to look at the source code of "nice" and "nohup" before re-inventing a squared wheel.

The second statement does not match with the design: it could be a bug to fix, it could be something is not used as expected. Let me explain: flom daemon recognizes a flom client abnormally terminates when it catches an error from the TCP/IP socket. The error might never arrive in some circumstances, but the TCP/IP stack provides a capabilities that should be activated using the SO_KEEPALIVE options. When I designed the behaviour, I avoided to invent my own polling technique and I relied on the standard TCP/IP stack. To make a long story short, the flom command line provides three parameters:

  --tcp-keepalive-time               Local override for SO_KEEPALIVE feature
  --tcp-keepalive-intvl               Local override for SO_KEEPALIVE feature
  --tcp-keepalive-probes           Local override for SO_KEEPALIVE feature

They can be used to set exactly the desired value for the socket on the server side.
A starting point for the usage of the three parameters is here: https://www.tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/
The FLoM man page contains a brief explanation of the meaning of every parameter.
Have you tried them, before trying to circumvent the problem with the signal forwarder?

In the meantime, I think your use case, integration of FLoM with Jenkins is very interesting: have you ever posted it in https://dzone.com/ or something equivalent? I think it could be useful for other users to know about your work.
Even managing your patch as a fork could be useful for other users: as explained above, I will work on this topic, but I will take my time.

**Comment by karelklic** [comment]
Thanks for such a detailed response. Now after reading your reply and testing the latest (unpatched) version of FLoM I realized my memory failed me when writing the bug report.

I wrote that the lock is not released when FLoM client is terminated. This is incorrect. The lock is indeed released, but the client's subprocess keeps running and keeps accessing the now-unprotected resource.

Regarding the signal redirection, for me, that is just a possible technical solution to the issue. I don't think signal redirection is more (or less) useful or expected than the current behavior.

Regarding TCP/IP options: so far I have used FLoM only locally with a named pipe.

It seems the right thing would be to close this bug report and try to be clearer with a new one.
